### PR TITLE
cast variable

### PR DESCRIPTION
**Pull request by astridx**
Pull Request for Issue in the [German forum](https://forum.joomla.de/thread/10945-unerw%C3%BCnschte-ausgabe-der-angaben-f%C3%BCr-text-und-wert-eines-optionsfelds-in-artikel/?postID=65285#post65285).

### Summary of Changes
Cast a default Custom Field value of `0` to `-0`.

### Testing Instructions
1. Create a Custom Field of type radio and set the default value to `0` but add no value with `0`.
Beiträge  Feld bearbeiten   test   Administration
2. Save the Custom Field.

### Expected result
You see an error if you want to save a custom field with a default value, that is not possible.

### Actual result
Saving is possible.

### Documentation Changes Required
No

### Note
I cast this value because the function empty is used twice: [Here](https://github.com/joomla/joomla-cms/blob/d3b544e92b04e6421cd683602e6535eb470ba276/administrator/components/com_fields/models/field.php#L220) and [here](https://github.com/joomla/joomla-cms/blob/d3b544e92b04e6421cd683602e6535eb470ba276/administrator/components/com_fields/models/field.php#L275) in the function `test`.

### CODE REVIEW

**Comment by N6REJ** [comment]
I have tested this item :white_check_mark: successfully on ecb7260380baf85b6c2da5d2ef84992501bdf8bc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27618">issues.joomla.org/tracker/joomla-cms/27618</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on ecb7260380baf85b6c2da5d2ef84992501bdf8bc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27618">issues.joomla.org/tracker/joomla-cms/27618</a>.</sub>

**Comment by Quy** [comment]
Not able to reproduce with v3.9.15.

**Comment by astridx** [comment]
Strange, I am able to reproduce with v3.9.15 today.

**Comment by Quy** [comment]
Sorry false alarm. I had applied your PR when testing. Duh! To me, it feels hackish fix.

**Comment by astridx** [comment]
You are right, this is hackish!
It would be more correct to change `empty()` to `isset()`.

The problem is that `emty("0")` is interpreted as `"there is no default value"`. Even `"0"` as a string! That is not correct. There is a default value in our case! 

Do you like my new version better?

**Comment by Quy** [comment]
Now it is always invalid.

**Comment by astridx** [comment]
@Quy The last change is embarrassing to me. 

Now it should fit.

The following things are considered to be empty (https://www.php.net/empty):

- "" (an empty string)
- 0 (0 as an integer)
- 0.0 (0 as a float)
- "0" (0 as a string)
- NULL
- FALSE
- array() (an empty array)
and
- var $var; (a variable declared, but without a value in a class)

If I am correct, we have to treat the case string "0" differently.

**Comment by astridx** [comment]
@N6REJ I would be happy if you could test again

File: administrator/components/com_fields/models/field.php [hunk 1]
```diff
@@ -217,7 +217,7 @@ public function save($data)
 	private function checkDefaultValue($data)
 	{
 		// Empty default values are correct
-		if (empty($data['default_value']))
+		if (empty($data['default_value']) && $data['default_value'] !== "0")
```
Comment (Quy) [inline administrator/components/com_fields/models/field.php]:
```suggestion
		if (empty($data['default_value']) && $data['default_value'] !== '0')
```

File: libraries/src/Form/Rule/OptionsRule.php [hunk 2]
```diff
@@ -42,7 +42,7 @@ public function test(\SimpleXMLElement $element, $value, $group = null, Registry
 		// Check if the field is required.
 		$required = ((string) $element['required'] == 'true' || (string) $element['required'] == 'required');
 
-		if (!$required && empty($value))
+		if (!$required && empty($value) && !($data['default_value'] !== "0"))
```
Comment (Quy) [inline libraries/src/Form/Rule/OptionsRule.php]:
```suggestion
		if (!$required && empty($value) && !($data['default_value'] !== '0'))
```

Comment (SharkyKZ) [inline libraries/src/Form/Rule/OptionsRule.php]: `$data` is undefined here.

Comment (astridx) [reply libraries/src/Form/Rule/OptionsRule.php]: Thanks, corrected

Comment (astridx) [reply libraries/src/Form/Rule/OptionsRule.php]: Done

**Comment by astridx** [comment]
Requested updates are done.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 8a61ac9a2c91fbaad7cff8ca1d042bcef62a7092<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27618">issues.joomla.org/tracker/joomla-cms/27618</a>.</sub>

**Comment by SharkyKZ** [comment]
Having a closer look, integer 0 and float 0.0 should also pass. So it would probably be better to compare against actually empty values (`null`, empty string and maybe empty array) rather than do `empty()` check. And we might need this for other validation rules as well.

**Comment by astridx** [comment]
>  @SharkyKZ  Having a closer look, integer 0 and float 0.0 should also pass. So it would probably be better to compare against actually empty values (`null`, empty string and maybe empty array) rather than do `empty()` check. 

I hope I understand you correctly. `0.0` does not pass as we use it as a string.

Articles  Edit Field   test   Administration(3)

Do you mean this?


> @SharkyKZ   And we might need this for other validation rules as well.

You're right. If I find a solution here that is good, I would look at more. I think it makes this PR more complicated if I try to includ this here.

**Comment by SharkyKZ** [comment]
This issue affects core form validation rules, so not just custom fields but any forms using them.

**Comment by astridx** [comment]
@SharkyKZ I do not understand what you mean. I know that this rule does not just apply to custom fields.
I prevent a string '0' from being considered as `no default value`. In my opinion, this is also correct in the core. Don't you see it that way? Can you give me an example of where this is wrong.

**Comment by ReLater** [comment]
I haven't tested it but what's about `is_numeric` ?
https://www.php.net/manual/en/function.is-numeric.php

**Comment by ReLater** [comment]
```
$tests = array(
-0,
0,
0.0,
.0,
'-0',
'0',
'0.0',
'.0',
'',
null
);

foreach ($tests as $element) {
 if (is_numeric($element)) {
  echo gettype($element)  . ':', PHP_EOL;
  echo var_export($element, true) . " is numeric" . PHP_EOL;
 } else {
  echo gettype($element)  . ':', PHP_EOL;
  echo var_export($element, true) . " is NOT numeric", PHP_EOL;
 }
}
exit;
```
```
integer:
0 is numeric
integer:
0 is numeric
double:
0.0 is numeric
double:
0.0 is numeric
string:
'-0' is numeric
string:
'0' is numeric
string:
'0.0' is numeric
string:
'.0' is numeric
string:
'' is NOT numeric
NULL:
NULL is NOT numeric
```
```
if (empty($val) && !is_numeric($val))
{
 --- DO sth. ---
}
```

**Comment by astridx** [comment]
OK. Now I saw it. `$value` is of type [`mixed`](https://www.php.net/manual/en/language.pseudo-types.php#language.types.mixed) - although I cannot find a place for thus `OptionsRule` that contains anything other than string.

But I think we can use the `empty()` function to make it easier. I just updated the branch.

What do you all mean?

**Comment by SharkyKZ** [comment]
This doesn't cover negative values. If you want to keep logic this way, use `is_numeric()` like suggested by @ReLater.

**Comment by astridx** [comment]
What is not covered?
Signed 0 (-0) is in PHP equal to not signed 0 for float and int: https://stackoverflow.com/questions/31250241/is-minus-negative-zero-equivalent-to-0-in-php?answertab=votes#tab-top
Am I missing something?

**Comment by SharkyKZ** [comment]
No, you're right. My bad. This works fine.

**Comment by ReLater** [comment]
#### This test is irrelevant! My fault.

I have tested this item :red_circle: unsuccessfully on 3ccf188abb9a4e0ae251d69c4bcb6fcf71615354<br /><br />Tested like instructed with today's nightly.

Radio Value: 1 => 1

Default Values:
0
0.0
.0
-0

"Save failed with the following error: The default value is invalid."<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27618">issues.joomla.org/tracker/joomla-cms/27618</a>.</sub>

**Comment by ReLater** [comment]
I have tested this item :white_check_mark: successfully on 3ccf188abb9a4e0ae251d69c4bcb6fcf71615354<br /><br />Ah sorry. There shall be the error if 0 entered as default value.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27618">issues.joomla.org/tracker/joomla-cms/27618</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 3ccf188abb9a4e0ae251d69c4bcb6fcf71615354<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27618">issues.joomla.org/tracker/joomla-cms/27618</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27618">issues.joomla.org/tracker/joomla-cms/27618</a>.</sub>

**Comment by HLeithner** [comment]
Thanks

**Pull request merged**